address_book/src/lib.rs

```diff
@@ -29,69 +35,131 @@ static LOGGING: Lazy<()> = Lazy::new(|| unsafe {
     tide::log::with_level(LOG_LEVEL);
 });
 
-trait Store {
-    fn save(&self, address: UserAddress, pubkey: UserPubKey) -> Result<(), std::io::Error>;
-    fn load(&self, address: UserAddress) -> Option<UserPubKey>;
+#[async_trait]
+pub trait Store: Clone + Send + Sync {
+    async fn save(&self, address: &UserAddress, pub_key: &UserPubKey)
+        -> Result<(), std::io::Error>;
+    async fn load(&self, address: &UserAddress) -> Option<UserPubKey>;
 }
 
 #[derive(Debug, Clone)]
-struct FileStore {
+pub struct FileStore {
     dir: PathBuf,
 }
 
+/// Persistent file backed store.
+/// Each (address, pub_key) pair is store in a single file inside `dir`.
 impl FileStore {
-    fn new(dir: PathBuf) -> Self {
+    pub fn new(dir: PathBuf) -> Self {
         Self { dir }
     }
-    fn path(&self, address: UserAddress) -> PathBuf {
+
+    fn path(&self, address: &UserAddress) -> PathBuf {
         let as_hex = hex::encode(bincode::serialize(&address).unwrap());
         self.dir.join(format!("{}.bin", as_hex))
     }
+
+    fn tmp_path(&self, address: &UserAddress) -> PathBuf {
+        let rand_string: String = rand::thread_rng()
+            .sample_iter(&Alphanumeric)
+            .take(16)
+            .map(char::from)
+            .collect();
+
+        self.path(address).with_extension(rand_string)
+    }
 }
 
+#[async_trait]
 impl Store for FileStore {
-    fn save(&self, address: UserAddress, pubkey: UserPubKey) -> Result<(), std::io::Error> {
-        fs::write(self.path(address), bincode::serialize(&pubkey).unwrap())
+    async fn save(
+        &self,
+        address: &UserAddress,
+        pub_key: &UserPubKey,
+    ) -> Result<(), std::io::Error> {
+        let tmp_path = self.tmp_path(address);
+        fs::write(&tmp_path, bincode::serialize(&pub_key).unwrap())?;
+        fs::rename(&tmp_path, self.path(address))
     }
 
-    fn load(&self, address: UserAddress) -> Option<UserPubKey> {
+    async fn load(&self, address: &UserAddress) -> Option<UserPubKey> {
         match fs::read(self.path(address)) {
             Ok(bytes) => Some(bincode::deserialize(&bytes).unwrap()),
             Err(_) => None,
         }
     }
 }
 
+/// Non-persistent in-memory store. Suitable for testing only.
```

Comment:
But why, though? Wouldn't using the same filesystem store code, but with a random tempdir that gets created on init and deleted on scope exit be better?

Reply:
Yes.

Reply:
https://github.com/EspressoSystems/cape/pull/718/commits/1038703b7866671b5b45f38fc4829645e93b85ef